tensorflow/core/kernels/cwise_ops.h

```diff
@@ -236,6 +236,110 @@ struct functor_traits<scalar_compose_op<Scalar, UnaryFunctor, BinaryFunctor>> {
   };
 };
 
+// tag for not complex number pass to tanh
+struct tanh_not_complex_tag {};
+// tag for complex number pass to tanh
+struct tanh_complex_tag {};
+// struct for tag dispatching
+template <typename T>
+struct tanh_dispatching_tag {
+  typedef tanh_not_complex_tag type;
+};
+// struct specify for complex<double>
+template <>
+struct tanh_dispatching_tag<std::complex<double>> {
+  typedef tanh_complex_tag type;
+};
+// struct specify for complex<float>
+template <>
+struct tanh_dispatching_tag<std::complex<float>> {
+  typedef tanh_complex_tag type;
+};
+// compute tanh for complex number
+template <typename T>
+const T tanh_dispatching(tanh_complex_tag, const T& _x) {
+  return numext::tanh(_x);
+}
+// compute tanh for not complex number
+template <typename T>
+const T tanh_dispatching(tanh_not_complex_tag, const T& _x) {
```

Comment:
I'll move it to bitbucket, and write two other struct for complex number, tag dispatching is complicated.